resolving/npm-resolver/src/pickPackageFromMeta.ts

```diff
@@ -140,3 +112,64 @@ export function pickVersionByVersionRange (
   }
   return maxVersion
 }
+
+function prioritizePreferredVersions (
+  meta: PackageMeta,
+  versionRange: string,
+  preferredVerSels?: VersionSelectors
+): string[][] {
+  const preferredVerSelsArr = Object.entries(preferredVerSels ?? {})
+  const versionsPrioritizer = new PreferredVersionsPrioritizer()
+  for (const [preferredSelector, preferredSelectorType] of preferredVerSelsArr) {
+    if (preferredSelector === versionRange) continue
+    switch (preferredSelectorType) {
+    case 'tag': {
+      versionsPrioritizer.add(meta['dist-tags'][preferredSelector])
+      break
+    }
+    case 'range': {
+      // This might be slow if there are many versions
+      // and the package is an indirect dependency many times in the project.
+      // If it will create noticeable slowdown, then might be a good idea to add some caching
+      const versions = Object.keys(meta.versions)
+      for (const version of versions) {
+        if (semver.satisfies(version, preferredSelector, true)) {
+          versionsPrioritizer.add(version)
+        }
+      }
+      break
+    }
+    case 'version': {
+      if (meta.versions[preferredSelector]) {
+        versionsPrioritizer.add(preferredSelector)
+      }
+      break
+    }
+    }
+  }
+  return versionsPrioritizer.versionsByPriority()
+}
+
+class PreferredVersionsPrioritizer {
+  private preferredVersions: Record<string, number> = {}
+
+  add (version: string) {
+    if (!this.preferredVersions[version]) {
+      this.preferredVersions[version] = 1
+    } else {
+      this.preferredVersions[version]++
+    }
+  }
+
+  versionsByPriority () {
+    const versionsByOccurrences = Object.entries(this.preferredVersions)
+      .reduce((acc, [version, occurrences]) => {
+        acc[occurrences] = acc[occurrences] ?? []
+        acc[occurrences].push(version)
+        return acc
+      }, {} as Record<number, string[]>)
+    return Object.keys(versionsByOccurrences)
+      .sort((a, b) => parseInt(b, 10) - parseInt(a, 10))
+      .map((occurrences) => versionsByOccurrences[parseInt(occurrences, 10)])
```

Comment:
nit: `occurrences` -> `occurrence` since this is a single number in this function.

Reply:
I think plural should be used here because it is the number of occurrences.

Reply:
Ah, no you're right. Disregard my comment.